pystac/extensions/file.py

```diff
@@ -0,0 +1,235 @@
+
+        self.item = item
+
+    def apply(self,
+              data_type=None,
+              size=None,
+              nodata=None,
+              checksum=None):
+        """Applies file extension properties to the extended Item.
+
+        Args:
+            data_type (FileDataType): The data type of the file.
+            size (int or None): size of the file in bytes.
+            nodata (List[Object] or None): Value(s) for no-data.
+            checksum (str or None): Multihash for the corresponding file,
+                encoded as hexadecimal (base 16) string with lowercase letters.
+        """
+        self.data_type = data_type
+        self.size = size
+        self.nodata = nodata
+        self.checksum = checksum
+
+    @property
+    def data_type(self):
+        """Get or sets the data_type of the file.
+
+        Returns:
+            FileDataType
+        """
+        return self.get_data_type()
+
+    @data_type.setter
+    def data_type(self, v):
+        self.set_data_type(v)
+
+    def get_data_type(self, asset=None):
+        """Gets an Item or an Asset data_type.
+
+        If an Asset is supplied and the data_type property exists on the Asset,
+        returns the Asset's value. Otherwise returns the Item's value
+
+        Returns:
+            FileDataType
+        """
+        if asset is not None and 'file:data_type' in asset.properties:
+            data_type = asset.properties.get('file:data_type')
+        else:
+            data_type = self.item.properties.get('file_data_type')
+
+        if data_type is not None:
+            return FileDataType(data_type)
+
+    def set_data_type(self, data_type, asset=None):
+        """Set an Item or an Asset data_type.
+
+        If an Asset is supplied, sets the property on the Asset.
+        Otherwise sets the Item's value.
+        """
+        if asset is not None:
+            asset.properties['file:data_type'] = data_type.value
+        else:
+            self.item.properties['file:data_type'] = data_type.value
+
+    @property
+    def size(self):
+        """Get or sets the size in bytes of the file
+
+        Returns:
+            int or None
+        """
+        return self.get_size()
+
+    @size.setter
+    def size(self, v):
+        self.set_size(v)
+
+    def get_size(self, asset=None):
+        """Gets an Item or an Asset file size.
+
+        If an Asset is supplied and the Item property exists on the Asset,
+        returns the Asset's value. Otherwise returns the Item's value
+
+        Returns:
+            float
+        """
+        if asset is None or 'file:size' not in asset.properties:
+            return self.item.properties.get('file:size')
+        else:
+            return asset.properties.get('file:size')
+
+    def set_size(self, size, asset=None):
+        """Set an Item or an Asset size.
+
+        If an Asset is supplied, sets the property on the Asset.
+        Otherwise sets the Item's value.
+        """
+        if asset is None:
```

Comment:
We should handle setting optional values to `None` by popping the dict keys - I've fixed an issue in https://github.com/stac-utils/pystac/pull/269/files that has a pattern for this which you could reuse here.

Reply:
@lossyrob Instead of replacing it in all the setters, I have extracted the mechanism to a separate method `_set_property`. Maybe it would be worth to have that in the parent Item class? Other extensions might have the same issue (I actually used the `eo` extension as the base for the `file ` one, so that one for sure has the same problem with None values in setters.

Let me know if you think it's a good idea and I can take care of that bit of refactoring

Reply:
Sounds like a great idea!

Reply:
Seems like I captured the idea of doing this in https://github.com/stac-utils/pystac/issues/230, so this refactor will close that issue.